packages/block-library/src/comment-template/edit.js

```diff
@@ -114,68 +114,73 @@ const CommentsList = ( {
 
 export default function CommentTemplateEdit( {
 	clientId,
-	context: { postId, 'comments/perPage': perPage, 'comments/order': order },
+	context: {
+		postId,
+		'comments/perPage': perPage,
+		'comments/order': order,
+		'comments/defaultPage': defaultPage,
+		'comments/inherit': inherit,
+	},
 } ) {
 	const blockProps = useBlockProps();
 
 	const [ activeComment, setActiveComment ] = useState();
-	const { commentOrder, commentsPerPage } = useSelect( ( select ) => {
-		const { getSettings } = select( blockEditorStore );
-		return getSettings().__experimentalDiscussionSettings;
+
+	const commentQuery = useCommentQueryArgs( {
+		postId,
+		perPage,
+		defaultPage,
+		inherit,
 	} );
-	const { rawComments, blocks } = useSelect(
+
+	const { topLevelComments, blocks } = useSelect(
 		( select ) => {
 			const { getEntityRecords } = select( coreStore );
 			const { getBlocks } = select( blockEditorStore );
 
-			const commentQuery = {
-				post: postId,
-				status: 'approve',
-				context: 'embed',
-				order: order || commentOrder,
-			};
-
-			if ( order ) {
-				commentQuery.order = order;
-			}
 			return {
-				rawComments: getEntityRecords(
-					'root',
-					'comment',
-					commentQuery
-				),
+				// Request only top-level comments. Replies are embedded.
+				topLevelComments: commentQuery
+					? getEntityRecords( 'root', 'comment', commentQuery )
+					: null,
 				blocks: getBlocks( clientId ),
 			};
 		},
-		[ postId, clientId, order ]
+		[ clientId, commentQuery ]
 	);
 
-	// TODO: Replicate the logic used on the server.
-	perPage = perPage || commentsPerPage;
-	// We convert the flat list of comments to tree.
-	// Then, we show only a maximum of `perPage` number of comments.
-	// This is because passing `per_page` to `getEntityRecords()` does not
-	// take into account nested comments.
-	const comments = useMemo(
-		() => convertToTree( rawComments ).slice( 0, perPage ),
-		[ rawComments, perPage ]
+	const { commentOrder } = useSelect( ( select ) => {
+		const { getSettings } = select( blockEditorStore );
+		return getSettings().__experimentalDiscussionSettings;
+	} );
+	order = order || commentOrder;
+
+	// Reverse the order of top comments if needed, as specified in the
+	// Discussion settings. NOTE that this only changes the order of comments in
+	// the given page, not the order of pages.
+
+	// Generate a tree structure of comment IDs.
+	const { commentTree } = useCommentTree(
+		order === 'desc'
+			? topLevelComments?.slice().reverse()
```

Comment:
Minor note, I think we typically prefer array spread over `.slice()` for copying:

```suggestion
			? [ ...topLevelComments ].reverse()
```

(Not that it's a hard-and-fast rule 😅 )

It would look a bit less nice if we have to take care of the `null` case, but maybe we can move the `commentTree` definition below the [early return](https://github.com/WordPress/gutenberg/pull/38187/files?diff=split&w=0#diff-885985bee90d7d51f079b99d59d8562cba40b4a6f64258e86fc81b0dc0f11b87R169-R175)?

Reply:
Yeah, sure! The only reason I opted for that syntax was that `topLevelComments` could not exist, but I can refactor the code to avoid that case. I prefer array spread as well. 😆